internal/observability/reserved_headers.go

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2024 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package observability
+
+import (
+	"go.uber.org/net/metrics"
+)
+
+type (
+	// ReservedHeaderMetrics is a collection of metrics for reserved headers.
+	ReservedHeaderMetrics struct {
+		strippedVec *metrics.CounterVector
+		errorVec    *metrics.CounterVector
+	}
+
+	// ReservedHeaderReporter is a wrapper for ReservedHeaderMetrics that has source and dest.
+	ReservedHeaderReporter struct {
+		m            *ReservedHeaderMetrics
+		source, dest string
+	}
+)
+
+// NewReserveHeaderMetrics creates a new ReservedHeaderMetrics.
+func NewReserveHeaderMetrics(scope *metrics.Scope, transport string) *ReservedHeaderMetrics {
+	return &ReservedHeaderMetrics{
+		strippedVec: registerReservedHeaderStripped(scope, transport),
+		errorVec:    registerReservedHeaderError(scope, transport),
+	}
+}
+
+// IncStripped increments the stripped metric.
+func (m *ReservedHeaderMetrics) IncStripped(source, dest string) {
+	incHeaderMetric(m.strippedVec, source, dest)
+}
+
+// IncError increments the error metric.
+func (m *ReservedHeaderMetrics) IncError(source, dest string) {
+	incHeaderMetric(m.errorVec, source, dest)
+}
```

Comment:
Answer to "why" is simple, DRY.

But it's interesting. I checked with -gcflags, incHeaderMetric is not inlined, but IncStripped is inlined. When I tried to copy incHeaderMetric content into IncStripped, IncStripped itself stopped being inlined. So it change nothing, we can leave it as is.
(Although I can't explain, what's wrong with this simple code and why it can't be inlined.)

Reply:
I checked, these calls are inlined by compiler, so we're fine in this regard. And as we're good from performance perspective, I would say, it's nice to have some ownership separation - edge doesn't know about internal structure of ReservedHeaderMetrics. Plus it may be useful one day anyway.